### PR TITLE
Chore/sql editor pagination of results

### PR DESCRIPTION
**Pull request by joshenlim**
Add a limit field for the SQL editor that will automatically limit the results (unless the query itself has a limit)
Number of results have been shifted below and will have a suffix (auto limit n rows) if the auto limit is kicked in

image

image

### CODE REVIEW

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx [hunk 1]
```diff
@@ -46,48 +63,70 @@ const UtilityActions = ({
...
+            <DropdownMenuItem
+              className="justify-between"
+              onClick={() => {
+                setIntellisenseEnabled(!intellisenseEnabled)
+                toast.success(
+                  `Successfully ${intellisenseEnabled ? 'disabled' : 'enabled'} intellisense. ${intellisenseEnabled ? 'Please refresh your browser for changes to take place.' : ''}`
+                )
+              }}
+            >
+              <p>Intellisense enabled</p>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:88]: remove p ?

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx [hunk 2]
```diff
@@ -46,48 +63,70 @@ const UtilityActions = ({
...
 
       <DropdownMenu>
         <DropdownMenuTrigger asChild>
-          <Button
-            type="text"
-            className="px-1"
-            icon={<Keyboard size={14} className="text-foreground-light" />}
-          />
+          <Button asChild type="outline" iconRight={<ChevronDown size={14} />}>
+            <span>{ROWS_PER_PAGE_OPTIONS.find((opt) => opt.value === snap.limit)?.label}</span>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:112]: span ?

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx [hunk 3]
```diff
@@ -46,48 +63,70 @@ const UtilityActions = ({
...
+      </div>
 
       <DropdownMenu>
         <DropdownMenuTrigger asChild>
-          <Button
-            type="text"
-            className="px-1"
-            icon={<Keyboard size={14} className="text-foreground-light" />}
-          />
+          <Button asChild type="outline" iconRight={<ChevronDown size={14} />}>
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:111]: Could be `default` ? That outline button looks a little nasty

Comment (MildTomato) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:111]: which is my fault..

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx [hunk 4]
```diff
@@ -111,19 +113,16 @@ const UtilityPanel = ({
...
-              `(${result.rows.length.toLocaleString()})`}
+            <span className="translate-y-[1px]">Results</span>
           </TabsTrigger_Shadcn_>
           <TabsTrigger_Shadcn_ className="py-3 text-xs" value="chart">
-            Chart
+            <span className="translate-y-[1px]">Chart</span>
           </TabsTrigger_Shadcn_>
+          {result?.rows && <ResultsDropdown id={id} />}
         </div>
         <div className="flex gap-1 h-full items-center">
```
Comment (MildTomato) [inline apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityPanel.tsx:125]: these classes might not do anything now

**Review by MildTomato — COMMENTED** [review]
All comments, non blocking :D

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx [hunk 5]
```diff
@@ -311,13 +311,30 @@ const SQLEditor = () => {
...
 
+        // Remove lines and whitespaces
+        const cleanedSql = sql.trim().replaceAll('\n', ' ').replaceAll(/\s+/g, ' ')
+        // Check if need to auto limit rows
+        const appendAutoLimit =
+          snap.limit > 0 &&
+          cleanedSql.toLowerCase().startsWith('select') &&
+          !cleanedSql.endsWith('limit') &&
+          !cleanedSql.endsWith('limit;') &&
+          !!!cleanedSql.match('limit [0-9]*[;]?$')
```
Comment (alaister) [inline apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx:322]: triple bang?!?!

File: apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx (hunk 2 above)
Comment (joshenlim) [reply apps/studio/components/interfaces/SQLEditor/UtilityPanel/UtilityActions.tsx:112]: its a workaround for the removeChild / insertBefore error that we see when users use Google Translate (or some third party extension) - although i'm genuinely not sure whether this is worth pursuing to keep doing as a practice tbh ([ref](https://github.com/supabase/supabase/pull/26580))

atm inclined to remove the span tbh

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx (hunk 5 above)
Comment (joshenlim) [reply apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx:322]: hahaha im wondering what would be a cleaner way to do this 😅 
wanted to check for falsey value of that 

maybe `(!!cleanedSql.match(...) === false)` might be better you reckon? for readability

**Comment by saltcod** [comment]
should switching this dropdown rerun the query? 

screenshot-2024-05-29-at-12 58 44

**Comment by joshenlim** [comment]
> should switching this dropdown rerun the query?
hmm I don't think so - would rather the user manually run the query again (at least thats how TablePlus does it, so am inclined to follow it for now). Its more explicit too which I prefer, cause if say the user ran `select * from table` that returns 200k rows, and changes the value to no limit, the UI performance will degrade then

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx (hunk 5 above)
Comment (joshenlim) [reply apps/studio/components/interfaces/SQLEditor/SQLEditor.tsx:322]: i'll change it to just 1 hahaha

**Review by alaister — COMMENTED** [review]
To throw another spanner in the works, something like `select generate_series(1,1000);;;` is actually a valid SQL query since it ignores any number of semicolons

**Comment by joshenlim** [comment]
I'm gonna opt to compress all those actions into a dropdown when the screen is small and the AI assistant is open like this - 
- I'd rather keep all "actions" in one place instead of several
- Those actions that i'm compressing into a dropdown are all "secondary" imo, they don't need a spot up front in the UI if the real estate doesnt support it
Screenshot 2024-05-30 at 14 43 36

**Comment by joshenlim** [comment]
> To throw another spanner in the works, something like select generate_series(1,1000);;; is actually a valid SQL query since it ignores any number of semicolons

okay added a fix and also unit tests on this piece of the logic - but just FYI atm i'm leaning towards the check for when to append the limit be more restrictive (i.e want to avoid adding the limit param too easily and end up causing a syntax error for the user instead). otherwise i think we'd need to install a full pg parser in the dashboard to properly do this, but I'm thinking what we have now fits most of the use case? wdyt though @alaister

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.utils.ts [hunk 6]
```diff
@@ -109,3 +109,43 @@ export const compareAsNewSnippet = (sqlDiff: ContentDiff) => {
...
+    !hasComments &&
+    !hasMultipleQueries &&
+    cleanedSql.toLowerCase().startsWith('select') &&
+    !cleanedSql.endsWith('limit') &&
+    !cleanedSql.endsWith('limit;') &&
+    !cleanedSql.match('limit [0-9]*[;]?$')
+  return { cleanedSql, appendAutoLimit }
+}
+
+export const prefixWithLimit = (sql: string, limit: number = 0) => {
```
Comment (alaister) [inline apps/studio/components/interfaces/SQLEditor/SQLEditor.utils.ts:143]: very much a nit but shouldn't this be called `suffixWithLimit`? 😅

**Review by alaister — APPROVED** [review]
LGTM!

File: apps/studio/components/interfaces/SQLEditor/SQLEditor.utils.ts (hunk 6 above)
Comment (joshenlim) [reply apps/studio/components/interfaces/SQLEditor/SQLEditor.utils.ts:143]: oh yes hahahaha that is correct

**Pull request merged**